cpp/src/arrow/array/array_nested.cc

```diff
@@ -827,6 +827,15 @@ Result<std::shared_ptr<Array>> MapArray::FromArraysInternal(
     return Status::Invalid("Map key and item arrays must be equal length");
   }
 
+  if (null_bitmap != nullptr && offsets->null_count() > 0) {
```

Comment:
`offsets` could have `null_count() == -1` (`kUnknownNullCount`) meaning that offsets might contain nulls that are not accounted for in this test.

The good news is all you have to do is replace all the `null_count()` tests here with `MayHaveNulls()` which would also fix this issue: https://github.com/apache/arrow/issues/38553

Reply:
Same issue is also present (and should be fixed) for `ListArrayFromArrays`

Reply:
Deserves a separate issue then?

Reply:
I think it would be good to fix both together in one PR (given both code snippets are exactly the same, for MapArray it was copied from ListArray)

Reply:
Then we can reuse https://github.com/apache/arrow/issues/38553, will add a note there.

Reply:
Done here: https://github.com/apache/arrow/pull/41957